### PR TITLE
Storage Library 2.0.0 release

### PR DESCRIPTION
**Pull request by hasonmsft**
ALL
* Updated storage service version to 2016-05-31. Fore more information, please see - https://msdn.microsoft.com/en-us/library/azure/dd894041.aspx
* Fixed the issue that `BatchOperation` doesn't support socket reuse for some node versions.
* Fixed the issue that `BatchOperation` request pool size is too big when the socket reuse is supported.
* Added empty headers to string-to-sign.
* For response has body and no `content-type` header, try to parse the body using xml format.
* Fixed the issue that retry filter will continuously retry for client error like `ETIMEDOUT`.
* Added support for client side timeout. In order to set the timeout per API, please use `options.clientRequestTimeoutInMs`. To set the default value for all requests made via a particular service, please use `{blob|queue|table}Service.defaultClientRequestTimeoutInMs`.

BLOB
* Added support for large block blob. 
* Added `publicAccessLevel` to `ContainerResult` for the APIs `listContainersSegmented` and `listContainersSegmentedWithPrefix`.
* When specifiying access condition `If-None-Match: *` for reading, it will always fail.
* Returned content MD5 for range gets Blobs.
* Fixed the issue that `useTransactionalMD5` didn't take effect for downloading a big blob.

QUEUE
* `createMessage` callback has been changed from `errorOrResponse` to `errorOrResult<QueueMessageResult>`.

FILE
* Returned content MD5 for range gets Files.
* Fixed the issue that `useTransactionalMD5` didn't take effect for downloading a big file.
* Added support for listing files and directories with prefix, refer to `FileService.listFilesAndDirectoriesSegmentedWithPrefix`.

TABLE
* Fixed the issue that response in incorrect for table batch operation when the error response item is not the first item in the responses.

### CODE REVIEW

File: lib/common/filters/exponentialretrypolicyfilter.js [hunk 1]
```diff
@@ -76,9 +76,8 @@ ExponentialRetryPolicyFilter.prototype.shouldRetry = function (statusCode, reque
   var incrementDelta = Math.pow(2, retryData.retryCount) - 1;
   var boundedRandDelta = this.retryInterval * 0.8 + Math.floor(Math.random() * (this.retryInterval * 1.2 - this.retryInterval * 0.8));
   incrementDelta *= boundedRandDelta;
-
+      
   retryData.retryInterval = Math.min(this.minRetryInterval + incrementDelta, this.maxRetryInterval);
-  retryData.retryable = retryData.retryCount ? retryData.retryCount < this.retryCount : true;
```
Comment (yaxia) [inline lib/common/filters/exponentialretrypolicyfilter.js]: If retry count is the hard bar of retry policies, can we just return false here when it exceeds?

File: lib/common/filters/retrypolicyfilter.js [hunk 2]
```diff
@@ -128,10 +129,19 @@ RetryPolicyFilter._handle = function (self, requestOptions, next) {
         // In this case, we should not retry the request.
         // If the output stream already get sent to server and get error back, 
         // we should NOT retry within the SDK as the stream data is not valid anymore if we retry directly.
-        if (!returnObject.outputStreamSent && returnObject.error && azureutil.objectIsNull(returnObject.retryable) &&
-             ((!azureutil.objectIsNull(returnObject.response) && retryInfo.retryable) || 
-               (returnObject.error.code === 'ECONNREFUSED' || returnObject.error.code === 'ETIMEDOUT' || returnObject.error.code === 'ESOCKETTIMEDOUT' || returnObject.error.code === 'ECONNRESET' || returnObject.error.code === 'EAI_AGAIN'))) {
-
+        if (
+            !returnObject.outputStreamSent && returnObject.error && azureutil.objectIsNull(returnObject.retryable) && notExceedMaxRetryCount &&
```
Comment (yaxia) [inline lib/common/filters/retrypolicyfilter.js]: It becomes more complicated. How about adding some comments to explain what scenarios do these conditions indicate?

Comment (yaxia) [inline lib/common/filters/retrypolicyfilter.js]: BTW, can we unify retryData, retryContext and retryInfo?

Comment (hasonmsft) [inline lib/common/filters/retrypolicyfilter.js]: As we talked offline, we'll try to refactor this part in next release.

Comment (hasonmsft) [inline lib/common/filters/retrypolicyfilter.js]: As we talked offline, we'll try to refactor this part in next release.

Comment (hasonmsft) [inline lib/common/filters/retrypolicyfilter.js]: we'll try to refactor this part in next release.

**Comment by hasonmsft** [comment]
Will resend against master branch

**Pull request closed without merging**